frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/upload-pvc-form/upload-pvc-form.tsx

```diff
@@ -167,26 +180,40 @@ export const UploadPVCForm: React.FC<UploadPVCFormProps> = (props) => {
     setAccessMode(event.currentTarget.value);
   };
 
-  const onlyPvcSCs = React.useCallback((sc: StorageClass) => !isObjectSC(sc), []);
-
-  const operatingSystems = getTemplateOperatingSystems(commonTemplates);
-
-  const handleOs = (osKey) => {
-    const operatingSystem = operatingSystems.find((newOs) => newOs.id === osKey);
+  const handleGoldenCheckbox = (checked) => {
+    setIsGolden(checked);
+    if (checked) {
+      const operatingSystem = operatingSystems.find((newOs) => newOs.id === os);
+      setNamespace(operatingSystem?.dataVolumeNamespace || TEMPLATE_VM_GOLDEN_OS_NAMESPACE);
+      if (pvcName && !os) {
+        setPvcName('');
+      } else {
+        setPvcName(operatingSystem?.dataVolumeName || os);
+      }
+    }
+    if (!checked) {
+      setNamespace(props.namespace);
+    }
+  };
 
-    setOs(osKey);
-    setPvcName(operatingSystem?.dataVolumeName || osKey);
-    setNamespace(operatingSystem?.dataVolumeNamespace || TEMPLATE_VM_GOLDEN_OS_NAMESPACE);
+  const handleOs = (newOs) => {
+    setOs(newOs);
+    const operatingSystem = operatingSystems.find((o) => o.id === newOs);
+    setPvcName(operatingSystem?.dataVolumeName || newOs);
   };
 
+  const onlyPvcSCs = React.useCallback((sc: StorageClass) => !isObjectSC(sc), []);
+
   React.useEffect(() => {
-    if (isGolden && pvcName && !os) {
-      setPvcName('');
-    }
-    if (!isGolden) {
-      setNamespace(props.namespace);
+    if (!isDisabled) {
+      // os url param might already have a golden image
+      const isGoldenExists = goldenPvcs?.find((pvc) => getName(pvc) === osParam);
+      if (!isGoldenExists) {
+        handleOs(osParam);
+      }
     }
-  }, [isGolden, os, props.namespace, pvcName]);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isDisabled]);
```

Comment:
it could be confusing to change this on isDisabled for new users. IMO it is better to stick to something like isLoading so we know we are doing it only once and not when some other action disables it.